src/Spice86.Core/Emulator/Devices/ExternalInput/PicEventQueue.cs

```diff
@@ -0,0 +1,325 @@
+        PicEntry? prevEntry = null;
+        int removedCount = 0;
+
+        while (entry != null) {
+            if (entry.Handler == handler && entry.Value == val) {
+                PicEntry? next = entry.Next;
+                if (prevEntry != null) {
+                    prevEntry.Next = next;
+                } else {
+                    _nextEntry = next;
+                }
+
+                entry.Next = _freeEntry;
+                _freeEntry = entry;
+                removedCount++;
+                entry = next;
+                continue;
+            }
+
+            prevEntry = entry;
+            entry = entry.Next;
+        }
+
+        if (removedCount > 0 && _logger.IsEnabled(LogEventLevel.Debug)) {
+            string handlerName = GetHandlerName(handler);
+            _logger.Debug(
+                "Removed {RemovedCount} PIC events for handler {Handler} with value {Value}",
+                removedCount,
+                handlerName,
+                val);
+        }
+    }
+
+    /// <summary>
+    ///     Removes all queued events matching the provided handler.
+    /// </summary>
+    /// <param name="handler">Handler to remove.</param>
+    public void RemoveEvents(PicEventHandler handler) {
+        PicEntry? entry = _nextEntry;
+        PicEntry? prevEntry = null;
+        int removedCount = 0;
+
+        while (entry != null) {
+            if (entry.Handler == handler) {
+                PicEntry? next = entry.Next;
+                if (prevEntry != null) {
+                    prevEntry.Next = next;
+                } else {
+                    _nextEntry = next;
+                }
+
+                entry.Next = _freeEntry;
+                _freeEntry = entry;
+                removedCount++;
+                entry = next;
+                continue;
+            }
+
+            prevEntry = entry;
+            entry = entry.Next;
+        }
+
+        if (removedCount > 0 && _logger.IsEnabled(LogEventLevel.Debug)) {
+            string handlerName = GetHandlerName(handler);
+            _logger.Debug(
+                "Removed {RemovedCount} PIC events for handler {Handler}",
+                removedCount,
+                handlerName);
+        }
+    }
+
+    /// <summary>
+    ///     Executes due events, updates cycle counters, and prepares the next wake-up point.
+    /// </summary>
+    /// <returns>
+    ///     <see langword="true" /> when work was performed or remains queued; otherwise <see langword="false" />.
+    /// </returns>
+    /// <remarks>
+    ///     Ensures the CPU cycle counters are normalized before servicing entries and recalculates the next wake-up time
+    ///     after draining ready handlers.
+    /// </remarks>
+    public bool RunQueue() {
+        _cpuState.CyclesLeft += _cpuState.Cycles;
+        _cpuState.Cycles = 0;
+        if (_cpuState.CyclesLeft <= 0) {
+            return false;
+        }
+
+        double indexNd = _cpuState.TickIndexNd;
```

Comment:
TickIndexNd => TickIndexInCycles?

This whole thing is very complex.
Time is measured in 3 ways as far as I understand:
- Miliseconds
- Cycles
- Index

Why not using a single unit for the events which would be "time in nanoseconds"?
In C# there is System.Diagnostics.Stopwatch.GetTimestamp (used in current implementation)

We could have a clock that gets the current time in nanoseconds from the cpu cycles / a periodic Stopwatch call to avoid having bad perfs and just call that here. This would separate the concerns of a time based priority queue and time counting.